pkg/storer/reserve.go

```diff
@@ -276,21 +281,39 @@ func (db *DB) removeChunk(ctx context.Context, address swarm.Address, batchID []
 
 func (db *DB) reserveCleanup(ctx context.Context) error {
 	dur := captureDuration(time.Now())
+	removed := 0
 	defer func() {
 		db.metrics.MethodCallsDuration.WithLabelValues("reserve", "cleanup").Observe(dur())
-	}()
+		db.metrics.ReserveCleanup.Add(float64(removed))
+		db.logger.Info("cleanup finished", "removed", removed, "duration", dur())
 
-	return db.reserve.IterateChunksItems(db.repo, 0, func(ci reserve.ChunkItem) (bool, error) {
+		db.lock.Lock(reserveUpdateLockKey)
+		defer db.lock.Unlock(reserveUpdateLockKey)
 
-		ok, err := db.batchstore.Exists(ci.BatchID)
-		if err != nil {
-			return false, err
+		if err := db.reserve.RecountSize(db.repo.IndexStore()); err != nil {
+			db.logger.Error(err, "recount reserve size")
 		}
 
-		if !ok {
+		db.metrics.ReserveSize.Set(float64(db.ReserveSize()))
+
+	}()
+
+	ids := map[string]struct{}{}
+
+	err := db.batchstore.Iterate(func(b *postage.Batch) (bool, error) {
+		ids[string(b.ID)] = struct{}{}
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return db.reserve.IterateChunksItems(db.repo, 0, func(ci reserve.ChunkItem) (bool, error) {
+		if _, ok := ids[string(ci.BatchID)]; !ok {
+			removed++
+			db.logger.Debug("cleanup expired batch", "batch_id", hex.EncodeToString(ci.BatchID))
```

Comment:
Logging might slow the iteration.

Reply:
we should only catch a few chunks here :)